xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp

```diff
@@ -162,6 +169,40 @@ bool CRendererBase::Configure(const VideoPicture& picture, float fps, unsigned o
   m_fps = fps;
   m_renderOrientation = orientation;
 
+  if (picture.hasDisplayMetadata || picture.color_primaries == AVCOL_PRI_BT2020)
```

Comment:
This is a static HDR metadata and in DXVAHD.cpp  is dynamic HDR metadata or HDR10+ / Doby Vision  (SEI packets).

Are a different things.

Reply:
I say nothing about DXVAHD. As I said early meta may be changed during playback so meta must be set on render. If dxva renderer isn't used you will never set "dynamic meta". isn't?

Reply:
Not understand you => "so meta must be set on render."

Meta is set on RenderBase.ccp -> Configure        This is not render?

If dxva renderer isn't used you will never set "dynamic meta". => Correct is needed dxva for set dynamic meta...  but dynamic meta is optional HDR10+ / Vision Only
HDR10 not needs this.  HDR10 only have STATIC meta and needs set at Configure.

Reply:
To clarificate:

Static meta = MediaInfo primaries, etc. --> Set with `IDXGISwapChain4::SetHDRMetaData` on Configure

Dynamic meta = stream SEI mesages --> Set with `ID3D11VideoContext2::VideoProcessorSetStreamHDRMetaData` on render

set static meta = requred for HDR10 / HDR10+ / Dolby Vision
set dynamic meta = optional for HDR10+ / Dolby Vision (really not required for HDR10)

Due a Microsoft bug or feature without call  `IDXGISwapChain4::SetHDRMetaData`  setting only dynamic meta in render does nothing.

Reply:
And:

If `VideoProcessorSetStreamHDRMetaData` is called 
also is needed `VideoProcessorSetOutputHDRMetaData` to pass monitor light parameters to render.

If not, image is rendered incorrectly.

Reply:
> Static meta = MediaInfo primaries, etc. --> Set with `IDXGISwapChain4::SetHDRMetaData` on Configure

this must be in Render method. How many times should I say this?


> Dynamic meta = stream SEI mesages --> Set with `ID3D11VideoContext2::VideoProcessorSetStreamHDRMetaData` on render

what is "dynamic meta"? why it processed only in DXVA renderer? why PS renderer is ignored?

Reply:
https://www.avforums.com/articles/what-is-4k-hdr-dynamic-metadata.13402

Reply:
What is static metadata?
A key part of this tone mapping is the metadata that is included with the content, the more information included the more accurate the tone mapping. When it comes to HDR there are two types of metadata, Static and Dynamic, with the former being standardised by SMPTE ST 2086 (Mastering Display Colour Volume).

This defines the static metadata that is supported by HDMI 2.0a, and is included with mastered HDR content to convey the colour volume of the mastering display and the luminance (brightness) of the content. This is described using the red, green, and blue primary colours and white point of the mastering display, plus its black level and peak luminance level. SMPTE ST 2086 also conveys the following luminance attributes of the mastered content:

MaxCLL (Maximum Content Light Level) – The MaxCLL is the luminance of the brightest pixel in the content expressed as nits;

Max FALL (Maximum Frame-Average Light Level) – The average luminance of all the pixels in each frame is first calculated and the MaxFALL is then the maximum value of the frame-average for all the frames in the content in nits.

The problem with static metadata is that if the tone mapping is performed without scene-by-scene content information, the mapping will be based only on the brightest scene and the widest gamut scene in the entire content. As a result the majority of the content will have greater compression of dynamic range and colour gamut than should really be necessary. So the less capable an HDR display is, the more important it is that the content is correctly tone mapped. That brings us on to dynamic metadata, which includes more information to help less capable displays tone map the content correctly.

Reply:
What is dynamic metadata?
Dynamic metadata allows a compatible display to tone map the HDR content to a smaller colour volume only as needed, when the content exceeds the capability of the playback display itself. The metadata can change dynamically, based on the minimum, maximum and average luminance and gamut requirements for each scene.

Tone mapping is more important the greater the difference between the mastering display and the playback display and will be an essential part of the future proofing of HDR technology. It ensures that playback displays that can do accurate tone mapping will still show content well, even when mastering displays are ultimately at or near the Rec. 2020 colour gamut limits and use a peak brightness of 10,000 nits.

Dynamic metadata has been standardised by SMPTE ST2094, which defines content-dependent metadata which will be supported in version 2.1 of the HDMI standard.

Reply:
What is SEI messages?
This annex specifies syntax and semantics for SEI message payloads.
SEI messages assist in processes related to decoding, display or other purposes. However, SEI messages are not required
for constructing the luma or chroma samples by the decoding process. Conforming decoders are not required to process
this information for output order conformance to this Specification (see Annex C and clause F.13 for the specification of
conformance). Some SEI message information is required to check bitstream conformance and for output timing decoder
conformance.
In clause C.5.2 and in clause F.13 including its subclauses, specification for presence of SEI messages are also satisfied
when those messages (or some subset of them) are conveyed to decoders (or to the HRD) by other means not specified in
this Specification. When present in the bitstream, SEI messages shall obey the syntax and semantics specified in
clause 7.3.5 and this annex. When the content of an SEI message is conveyed for the application by some means other than
presence within the bitstream, the representation of the content of the SEI message is not required to use the same syntax
specified in this annex. For the purpose of counting bits, only the appropriate bits that are actually present in the bitstream
are counted.

D.3.28 Mastering display colour volume SEI message semantics
This SEI message identifies the colour volume (the colour primaries, white point and luminance range) of a display
considered to be the mastering display for the associated video content – e.g., the colour volume of a display that was used
for viewing while authoring the video content. The described mastering display is a three-colour additive display system
that has been configured to use the indicated mastering colour volume.
This SEI message does not specify the measurement methodologies and procedures used for determining the indicated
values or any description of the mastering environment. It also does not provide information on colour transformations
that would be appropriate to preserve creative intent on displays with colour volumes different from that of the described
mastering display.
The information conveyed in this SEI message is intended to be adequate for purposes corresponding to the use of SMPTE
ST 2086.
When a mastering display colour volume SEI message is present for any picture of a CLVS of a particular layer, a mastering
display colour volume SEI message shall be present for the first picture of the CLVS. The mastering display colour volume
SEI message persists for the current layer in decoding order from the current picture until the end of the CLVS. All
mastering display colour volume SEI messages that apply to the same CLVS shall have the same content.
display_primaries_x[ c ] and display_primaries_y[ c ] specify the normalized x and y chromaticity coordinates,
respectively, of the colour primary component c of the mastering display in increments of 0.00002, according to the CIE
1931 definition of x and y as specified in ISO 11664-1 (see also ISO 11664-3 and CIE 15). For describing mastering
displays that use red, green and blue colour primaries, it is suggested that index value c equal to 0 should correspond to
the green primary, c equal to 1 should correspond to the blue primary and c equal to 2 should correspond to the red colour
primary (see also Annex E and Table E.3). The values of display_primaries_x[ c ] and display_primaries_y[ c ] shall be in
the range of 0 to 50 000, inclusive.
white_point_x and white_point_y specify the normalized x and y chromaticity coordinates, respectively, of the white
point of the mastering display in normalized increments of 0.00002, according to the CIE 1931 definition of x and y as
specified in ISO 11664-1 (see also ISO 11664-3 and CIE 15). The values of white_point_x and white_point_y shall be in
the range of 0 to 50 000.
max_display_mastering_luminance and min_display_mastering_luminance specify the nominal maximum and
minimum display luminance, respectively, of the mastering display in units of 0.0001 candelas per square metre.
min_display_mastering_luminance shall be less than max_display_mastering_luminance.
At minimum luminance, the mastering display is considered to have the same nominal chromaticity as the white point.

Reply:
> why it processed only in DXVA renderer? why PS renderer is ignored?

Dynamic metadata is the future but actually does nothing. 
This is why even with the entire code block commented on DXVAHD.ccp it work OK in the first version.

Actually only setting Static metadata at playing begin `IDXGISwapChain4::SetHDRMetaData` is sufficient.

Reply:
A lot of text especially about SEI which unhelpful at all for this.

I will try to help you to understand things. 

1. ffmpeg parses SEI messages to extract HDR10[+] metadata, static and dynamic. right? right.
2. ffmpeg stores metas (static and dynamic) into side data of a frame (each frame). right? right.
3. where is static meta stored by ffmpeg? in side data with `AV_FRAME_DATA_MASTERING_DISPLAY_METADATA` flag see https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/mastering_display_metadata.c#L34
4. where is dynamic meta stored by ffmpeg? in side data with `AV_FRAME_DATA_DYNAMIC_HDR_PLUS` flag see https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hdr_dynamic_metadata.c#L38
5. Is `displayMetadata` of videoPicture contains static or dynamic meta? It contains only **static** meta. because `DVDVideoCodecFFmpeg.cpp` doesn't process `AV_FRAME_DATA_DYNAMIC_HDR_PLUS` side data see https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecFFmpeg.cpp#L1000
6. Can static meta be changed during playback? yes it can. 

So I still have a couple of questions:
1. Why you set meta of first frame for the whole video?
2. Why you decided that `DXVAHD` processes (or can process) dynamic data if dynamic data is absent in `videoPicture` at all?

Reply:
1. Why you set meta of first frame for the whole video?

Because usually (100% in HDR10) this info (Static meta) no change for whole video. Because this is called **Static** meta. Also HDR10 does **not** support Dynamic meta so dynamic meta = static meta = static.

Even HDR10+ needs to be retrocompatible with "old" HW and SW so also includes static meta.
In this case Dynamic meta change but Static not.

Static meta is set with `IDXGISwapChain4::SetHDRMetaData` and can be picked from media container headers (MediaInfo), not transport stream SEI messages.


2. Why you decided that DXVAHD processes (or can process) dynamic data if dynamic data is absent in videoPicture at all?

Because why **should** I do it. Probably, and you confirmed now, this is not implemented currently but it depends on lower layers (ffmpeg - DVDVideoCodecFFmpeg) out of this scope. 

When DVDVideoCodecFFmpeg returns dynamic meta can be handled in render with minimal or no changes. (Static meta is not need process for every frame so not needed in render routine).

As I said in past messages everything works just as well for HDR10 if metadata is not processed in the render routine. Also for now Microsoft does neither support PLUS data extension in `VideoProcessorSetStreamHDRMetaData`

https://docs.microsoft.com/en-us/windows/win32/api/d3d11_4/nf-d3d11_4-id3d11videocontext2-videoprocessorsetstreamhdrmetadata


But the most important of all is that: as it is now **it works** in a stable way. Tested with multiple media samples HDR, HDR10+ / Dolby Vision. (if exist dynamic meta is ignored but image is displayed correctly since they are backward compatible systems).

https://forum.kodi.tv/showthread.php?tid=349861

Reply:
My question is: What changes would you make right now so that it works better?

Reply:
I recognize that I was wrong in some concept (due to vocabulary issues).

Let's simplify things:

1. Forget Dynamic HDR metadata (there is nothing to do with this at the moment).
2. Static HDR metadata can change in middle of stream. I agree.

What's wrong with using `IDXGISwapChain4::SetHDRMetaData` at Configure?

And re-set HDR metadata in render with: 
`ID3D11VideoContext2::VideoProcessorSetStreamHDRMetaData`

If the metadata changes, it is handled evenly.

Reply:
Microsoft says:

IDXGISwapChain4::SetHDRMetaData method
This method sets High Dynamic Range (HDR) and Wide Color Gamut (WCG) **header** metadata.

ID3D11VideoContext2::VideoProcessorSetStreamHDRMetaData method

Sets the HDR metadata associated with the video **stream**.

Reply:
I'm "work in progress" with add `SetHdrMetaData` in `CRendererBase::Render` 

...finally  :) 

However not have much time weekday...

Reply:
> What's wrong with using `IDXGISwapChain4::SetHDRMetaData` at Configure?

because:
1. it's called once on renderer creation and can be used for several videos with the same format.
2. meta can be absent for the first frame (because of ffmpeg or container issue or if media is corrupt)
3. meta can be changed.

Reply:
Done in https://github.com/xbmc/xbmc/pull/17012/commits/200fd8afe4447764b7b8c5a3c6424ab94d4dcc87